### PR TITLE
DEPS: Bump cython 3.0

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by WillAyd** [comment]
Looks pretty good. Are all of these changes required to make Cython 3.0 work or are some of the signature changes general cleanups that made sense to bundle?

@Dr-Irv FYI in case this impacts pandas-stubs

**Comment by lithomas1** [comment]
> Looks pretty good. Are all of these changes required to make Cython 3.0 work or are some of the signature changes general cleanups that made sense to bundle?
> 
> @Dr-Irv FYI in case this impacts pandas-stubs

Yeah, it turns out our stubs are horrendously outdated. I guess Cython 3.0 exposes the annotation information, so stubtest is complaining now.

**Comment by hkad98** [comment]
Do we really want to include this bump in the 2.1.0 release? AFAIK Apache Arrow will adapt Cython 3.0.0 in their 14.0.0, but they do not release 13.0.0 yet, meaning the integration with Apache Arrow will not work, right?

**Comment by lithomas1** [comment]
> Do we really want to include this bump in the 2.1.0 release? AFAIK Apache Arrow will adapt Cython 3.0.0 in their 14.0.0, but they do not release 13.0.0 yet, meaning the integration with Apache Arrow will not work, right?

We don't compile against Apache Arrow (only numpy). All integration with arrow is done through Python IIRC.

File: pandas/_libs/arrays.pyi [hunk 1]
```diff
@@ -26,7 +26,7 @@ class NDArrayBacked:
     def size(self) -> int: ...
     @property
     def nbytes(self) -> int: ...
-    def copy(self): ...
+    def copy(self, order: str = "C"): ...
```
Comment (jbrockmendel) [inline pandas/_libs/arrays.pyi:29]: any reason this is `"C"` instead of `...`?

Comment (Dr-Irv) [reply pandas/_libs/arrays.pyi:29]: In a PYI file, you don't specify default values.  Defaults would only go in the declaration of the function.  So it should be `order: str = ...` .  That means the `order` argument is optional.

File: pandas/_libs/tslibs/period.pyx [hunk 2]
```diff
@@ -2511,7 +2511,7 @@ cdef class _Period(PeriodMixin):
         object_state = None, self.freq, self.ordinal
         return (Period, object_state)
 
-    def strftime(self, fmt: str) -> str:
+    def strftime(self, fmt: str | None) -> str:
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/period.pyx:2514]: good one 👍

**Review by MarcoGorelli — APPROVED** [review]
Nice!

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Nice thanks @lithomas1

**Comment by jbrockmendel** [comment]
very cool to see this in!

**Pull request merged**